Refactor BEdita shell + solve problematic behaviour with SQLite

fquffio opened the pull request:
This PR fixes #1178.

The fix includes a major refactor of BEdita shell: each step in the `bin/cake bedita setup` command has been split into its own Task, that is unit tested separately. This helps with the problem that we used to experience, since we have more control on each step tests, and we're able to better handle the mess those tasks do with connections and database tables. Also, tasks are now more thoroughly tested.

fquffio committed:
#1178 refactor: refactor existing components to make them fit the rest

fquffio committed:
#1178 feat: implement shell task to setup connection (with wizard)

fquffio committed:
#1178 feat: implement shell task to check filesystem permissions

fquffio committed:
[minor] refactor: make help texts more consistent across tasks

fquffio committed:
#1178 feat: implement shell task to setup admin user

fquffio committed:
#1178 feat: add non-interactive connection setup

fquffio committed:
[minor] test: fix required fixtures for `ResourceShellTest`

fquffio committed:
#1178 test: update assertions for `CheckSchemaTaskTest`

fquffio committed:
#1178 fix: use transaction when dropping tables to keep schema consistent

fquffio committed:
#1178 refactor: refactor `bedita` shell to use tasks

fquffio committed:
[minor] refactor: use argument unpacking vs. `call_user_func_array()`

fquffio committed:
#1178 fix: several fixes to address CI failures

fquffio committed:
Merge remote-tracking branch 'origin/4-cactus' into issue/1178-fixtures-mess

# Conflicts:
#	plugins/BEdita/Core/src/Shell/BeditaShell.php

fquffio committed:
[minor] docs: add `@method` annotations to `ApplicationsTable`

fquffio committed:
#1178 refactor: split `CheckFilesystemTask::main()` for readability

fquffio committed:
#1178 refactor: add `CheckApiKeyTask`

fquffio committed:
[minor] test: fix coverage annotations

fquffio commented:
There's something weird with code coverage… For instance, [SetupConnectionTask](https://codecov.io/gh/bedita/bedita/pull/1371/tree/plugins/BEdita/Core/src/Shell/Task) seem to be not covered at all, while tests are present and `@covers` annotations have been set up properly, as far as I can tell… 🤔

stefanorosanelli commented:
Strange behavior indeed.

With `vendor/bin/phpunit --coverage-html ...` we get 0% coverage on [SetupConnectionTask](https://codecov.io/gh/bedita/bedita/pull/1371/tree/plugins/BEdita/Core/src/Shell/Task) instead with `vendor/bin/phpunit --coverage-html ... plugins/BEdita/Core/tests/TestCase/Shell/Task/` we get:

- 88.78% coverage on MySQL 
- 74.49% coverage on SQlite (less lines are covered in this case) but 2 failures...

stefanorosanelli commented on plugins/BEdita/Core/src/Shell/Task/CheckApiKeyTask.php:
```diff
@@ -0,0 +1,75 @@
+            $this->verbose('<error>FAIL</error>');
+            $this->abort('Default application is missing, please check your installation');
+
+            return false;
```
unreachable statement after `$this->abort()`

stefanorosanelli commented on plugins/BEdita/Core/src/Shell/Task/CheckFilesystemTask.php:
```diff
@@ -0,0 +1,160 @@
+        // Detect HTTP daemon user.
+        $httpdUser = $this->getHttpdUser();
+        if (empty($httpdUser)) {
+            $this->out('=====> <warning>Unable to detect webserver user</warning>');
```
missing coverage of empty `$httpdUser`, but we can live without :wink:

fquffio replied:
Actually, this part is covered by `CheckFilesystemTaskTest::testExecuteAutodetectOk()` and `CheckFilesystemTaskTest::testExecuteAutodetectFail()`. However, one of this test is always skipped (if a webserver is running, fail is skipped, if a webserver is _not_ running, ok is skipped). In local environment it is easy to test both cases by running tests then turning off Apache or whatever, and re-running them again.

stefanorosanelli replied:
Ok, I only looked at the coverage results...

fquffio committed:
[minor] style remove unreachable statement

stefanorosanelli commented:
Code coverage problem is somehow still related to test execution order: renaming `BEditaShellTest` to `ZBEditaShellTest` increases coverage of `SetupConnectionTask ` from 0% to 88.78%

fquffio commented:
@stefanorosanelli I'm trying to user more specific `@covers` annotations, to see if it solves the problem. As far as I can tell there's nothing wrong in our configuration, however… 🤔

fquffio committed:
[minor] test: use specific `@covers` annotations, instead of whole class

fquffio committed:
[minor] fix: don't import `SetupConnectionTaskTest` in `BeditaShellTest`

`BeditaShellTest` used to reference a constant defined in `SetupConnectionTaskTest`.
This caused the latter class to be autoloaded, which caused the `SetupConnectionTaskTest` case not
to be loaded by PHPUnit as per sebastianbergmann/phpunit#2136.

fquffio commented:
Ok, we ran into what looks to me like a PHPUnit bug, but might also be a kind of bad practice I'm not aware of. `BeditaShellTest` used to reference two constants defined in `SetupConnectionTaskTest`, thus that test case was autoloaded before being executed, and PHPUnit silently skipped it. c28153c should have worked arount this, and hopefully PR coverage will become a bit higher.

fquffio committed:
[minor] fix: fix tests if test connection has empty password

fquffio committed:
[minor] test: add missing `@covers` annotations

fquffio committed:
#1178 fix: escape all values when updating config

fquffio committed:
#1178 test: add test case to cover PHP syntax validation failure

fquffio committed:
#1178 fix: read stdout & stderr pipes to make exit code work as expected

stefanorosanelli commented on plugins/BEdita/Core/src/Shell/Task/SetupConnectionTask.php:
```diff
@@ -0,0 +1,315 @@
+        $connectionName = $this->param('connection');
+        $connection = ConnectionManager::get($connectionName);
+        if (!($connection instanceof Connection)) {
+            $this->abort('Invalid connection object');
```
[minor] is this check really necessary? 
we have it in many places, and is also a never tested case...

fquffio replied:
Mmh… Despite not strictly necessary right now, it might become useful in the future. `ConnectionManager::get()` is guaranteed to return a `ConnectionInterface` instance. Database connections are just one possible implementation of `ConnectionInterface`. For example, ElasticSearch connections will be different implementations of the same interface.

This is why I would leave it in a place where we are explicitly managing a database connection: it might cause issues in the future if a user runs `bin/cake bedita setup --connection elasticsearch`. I have added a test case to test this condition.

fquffio committed:
[minor] test: add test case to cover invocation with invalid connection
